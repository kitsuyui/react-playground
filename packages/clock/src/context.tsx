import React from 'react'
import { useState } from 'react'
import { useInterval } from 'react-use'

export interface ClockContextProviderProps {
  children: React.ReactNode
  refreshInterval?: number
}

export type ClockContextValue = Date

export const ClockContext = React.createContext<ClockContextValue>(new Date())

export const ClockContextProvider: React.FC<ClockContextProviderProps> = (
  props
): React.JSX.Element => {
  const { children } = props
  const [date, setDate] = useState(new Date())
  const refreshInterval = props.refreshInterval || 10 // default 10ms

  // use useInterval instead of setInterval
  // https://github.com/streamich/react-use/blob/master/docs/useInterval.md
  useInterval(() => {
    setDate(new Date())
  }, refreshInterval)

  return <ClockContext.Provider value={date}>{children}</ClockContext.Provider>
}
