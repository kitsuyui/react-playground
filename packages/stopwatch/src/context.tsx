import React from 'react'
import { useInterval } from 'react-use'
import { calcElapsedTime } from './time'

export interface StopwatchValue {
  elapsedTime: number
  running: boolean
}

export interface StopwatchActions {
  start(): void
  stop(): void
  toggle(): void
  reset(): void
}

export type StopwatchContextValue = StopwatchValue & StopwatchActions

// TODO: Support leap second: https://github.com/kitsuyui/react-playground/issues/40
export const StopwatchContext = React.createContext<StopwatchContextValue>({
  elapsedTime: 0,
  running: false,
  start: () => { /* do nothing */ },
  stop: () => { /* do nothing */ },
  toggle: () => { /* do nothing */ },
  reset: () => { /* do nothing */ },
})

export interface StopwatchProviderProps {
  refreshInterval?: number
  children: React.ReactNode
  onStart?(event: CustomEvent): void
  onStop?(event: CustomEvent): void
  onReset?(event: CustomEvent): void
}

export const StopwatchContextProvider: React.FC<StopwatchProviderProps> = (
  props
): React.JSX.Element => {
  const { children } = props
  const emptyFn = (_event: CustomEvent) => { /* do nothing */ }
  const onStart = props.onStart ?? emptyFn
  const onStop = props.onStop ?? emptyFn
  const onReset = props.onReset ?? emptyFn
  const [running, setRunning] = React.useState(false)
  const [elapsedTimeInLap, setElapsedTimeInLap] = React.useState(0)
  const [elapsedTimeTotal, setElapsedTimeTotal] = React.useState(0)
  const [startTime, setStartTime] = React.useState(new Date())

  const refreshInterval = props.refreshInterval || 10 // default 10ms

  useInterval(() => {
    if (running) {
      updateElapsedTime()
    }
  }, refreshInterval)

  const updateElapsedTime = () => {
    setElapsedTimeInLap(calcElapsedTime(startTime))
  }

  const moveLapToTotal = () => {
    setElapsedTimeTotal(elapsedTimeTotal + elapsedTimeInLap)
    setElapsedTimeInLap(0)
  }

  const start = () => {
    if (running) {
      return
    }
    onStart(new CustomEvent('start', {}))
    setStartTime(new Date())
    setRunning(true)
  }

  const stop = () => {
    if (!running) {
      return
    }
    onStop(new CustomEvent('stop', {}))
    updateElapsedTime()
    setRunning(false)
    moveLapToTotal()
  }

  const toggle = () => {
    if (running) {
      stop()
    } else {
      start()
    }
  }

  const reset = () => {
    if (running) {
      stop()
    }
    onReset(new CustomEvent('reset', {}))
    setElapsedTimeInLap(0)
    setElapsedTimeTotal(0)
  }

  const elapsedTime = elapsedTimeInLap + elapsedTimeTotal

  return (
    <StopwatchContext.Provider
      value={{
        elapsedTime,
        running,
        toggle,
        reset,
        start,
        stop,
      }}
    >
      {children}
    </StopwatchContext.Provider>
  )
}
