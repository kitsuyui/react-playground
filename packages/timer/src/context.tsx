import React from 'react'
import { useInterval } from 'react-use'

import { calcRemaining } from './time'

export interface TimerValue {
  remaining: number
  running: boolean
}

export interface TimerActions {
  start(): void
  stop(): void
  toggle(): void
  reset(): void
  incrementTimerValue(value: number): void
  setTimerValue(value: number): void
}

export type TimerContextValue = TimerValue & TimerActions

// TODO: Support leap second: https://github.com/kitsuyui/react-playground/issues/40
export const TimerContext = React.createContext<TimerContextValue>({
  remaining: 0,
  running: false,
  start: () => { /* do nothing */ },
  stop: () => { /* do nothing */ },
  toggle: () => { /* do nothing */ },
  reset: () => { /* do nothing */ },
  incrementTimerValue: (_value: number) => { /* do nothing */ },
  setTimerValue: (_value: number) => { /* do nothing */ },
})

export interface TimerContextProviderProps {
  refreshInterval?: number
  children: React.ReactNode
  onStart?(event: CustomEvent): void
  onStop?(event: CustomEvent): void
  onComplete?(event: CustomEvent): void
  onReset?(event: CustomEvent): void
}

export const TimerContextProvider: React.FC<TimerContextProviderProps> = (
  props
): React.JSX.Element => {
  const { children } = props
  const emptyFn = (_event: CustomEvent) => { /* do nothing */ }
  const onStart = props.onStart ?? emptyFn
  const onStop = props.onStop ?? emptyFn
  const onComplete = props.onComplete ?? emptyFn
  const onReset = props.onReset ?? emptyFn
  const [running, setRunning] = React.useState(false)
  const [targetDate, setTargetDate] = React.useState(new Date())
  const [remaining, setRemaining] = React.useState(0)

  const refreshInterval = props.refreshInterval || 10 // default 10ms

  useInterval(() => {
    tick()
  }, refreshInterval)

  const tick = () => {
    // It can't be ticked if it is not running.
    if (running) {
      const remaining = calcRemaining(targetDate)
      setRemaining(remaining)
      if (remaining <= 0) {
        setRemaining(0)
        setRunning(false)
        onComplete(new CustomEvent('complete', {}))
      }
    }
  }

  const toggle = () => {
    if (running) {
      stop()
    } else {
      start()
    }
  }

  const reset = () => {
    // First, stop the timer if it is running.
    if (running) {
      stop()
    }
    setRemaining(0)
    onReset(new CustomEvent('reset', {}))
  }

  const start = () => {
    // It can't be started if the remaining time is 0.
    if (remaining <= 0) {
      return
    }
    // It can't be started if it is already running.
    if (running) {
      return
    }
    setTargetDate(new Date(Date.now() + remaining * 1000))
    setRunning(true)
    onStart(new CustomEvent('start', {}))
  }

  const stop = () => {
    // It can't be stopped if it is not running.
    if (!running) {
      return
    }
    setRunning(false)
    onStop(new CustomEvent('stop', {}))
  }

  return (
    <TimerContext.Provider
      value={{
        remaining,
        running,
        toggle,
        reset,
        start,
        stop,
        setTimerValue(value: number) {
          stop()
          setRemaining(value || 0)
        },
        incrementTimerValue(value: number) {
          stop()
          setRemaining(remaining + value)
        },
      }}
    >
      {children}
    </TimerContext.Provider>
  )
}
