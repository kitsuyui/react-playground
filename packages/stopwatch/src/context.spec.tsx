import { describe, expect, it, } from 'vitest'

import { render } from '@testing-library/react'

import { StopwatchContextProvider, StopwatchContext } from './context'
import React from 'react'


describe('StopwatchContextProvider', () => {
  it('provides context', () => {
    const ContextReceiver = () => {
      const context = React.useContext(StopwatchContext)
      const valueString = JSON.stringify(Object.keys(context).sort())
      return (<>{valueString}</>)
    }
    const { asFragment } = render(<StopwatchContextProvider>
      <ContextReceiver />
    </StopwatchContextProvider>)
    expect(asFragment()).toMatchSnapshot()
  })
})
