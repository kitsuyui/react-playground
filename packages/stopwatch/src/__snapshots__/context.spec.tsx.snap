// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StopwatchContextProvider > provides context 1`] = `
<DocumentFragment>
  ["elapsedTime","reset","running","start","stop","toggle"]
</DocumentFragment>
`;
